cluster-image-builder/serve/_replica_scheduler/chwbl_scheduler.py

```diff
@@ -295,4 +296,69 @@ def replica_queue_len_cache(self) -> ReplicaQueueLengthCache:
     @property
     def curr_replicas(self) -> Dict[ReplicaID, RunningReplica]:
         """Return the current replicas."""
-        return self._replicas
\ No newline at end of file
+        return self._replicas
+
+    def _extract_cache_key(self, payload, request_id: str) -> str:
+        """Extract cache key from OpenAI-compatible chat completions payload.
+        
+        For chat completions, we want to hash based on:
+        1. System prompt (if present)
+        2. First N user messages (configurable)
+        
+        This ensures that similar conversation contexts are routed to the same replica.
+        """
+        # Default fallback
+        if not payload:
+            logger.info(f"No payload found, using request_id: {request_id}")
+            return str(request_id)
+        
+        try:
+            # Handle different payload formats (args could be tuple, list, or dict)
+            if isinstance(payload, (tuple, list)) and len(payload) > 0:
+                # Assume first argument contains the request data
+                request_data = payload[0]
+            elif isinstance(payload, dict):
+                request_data = payload
+            else:
+                # Fallback to string representation
+                return str(payload)
+            
+            # Extract relevant fields for cache key
+            cache_components = []
+            
+            # System prompt and user messages
+            messages = request_data.get('messages', [])
+            system_prompt = None
+            user_messages = []
+            
+            for msg in messages:
+                if isinstance(msg, dict):
+                    role = msg.get('role', '')
+                    content = msg.get('content', '')
+                    
+                    if role == 'system':
+                        system_prompt = content
+                    elif role == 'user':
+                        user_messages.append(content)
```

Comment:
when we get `_max_user_messages_for_cache` number of user messages, can we exit the loop early to avoid traversing all messages ?

Reply:
good catch, updated